### PR TITLE
2688 tanf section 2 validation clean up

### LINKED ISSUE #2688: TANF Section 2 validation clean-up
**Description:**
After #2633 merges, some of the validation checks for TANF Section 2 data needs updates. These include space-filling for some non-required data elements

[workbook for TANF section 2](https://gorafttech.sharepoint.com/:f:/r/sites/TDRSResearchDesign/Shared%20Documents/Dev/dev/Parsing/data/tanf/Section2?csf=1&web=1&e=dywxgb) updated. See column I in relevant worksheets for data elements that should allow for spaces

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] relevant T4 data elements allow for space-filling
- [ ] relevant T5 data elements allow for space-filling
- [ ] REC_SUB_HOUSING range is 1-2
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] update tanf section 2 (ssp and tribal section 2 have already been completed)
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Other cleanups in section 2 that are now captured in other tickets_ 

- ~`DATE_OF_BIRTH` in the T4  record are stored and validated as integers instead of date (YYYYMMDD)~
    - captured by #2746
- ~`if (calculated) AGE >= 19, then item 19A == 1 or 2` not implemented~
    - dob calculation changed by #2746, drop this req
- ~`FIPS_CODE` not stored in the T4 record~
    - captured by #2796

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

- _Doc 1_
- _Doc 2_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- do any updates need to be made to pre/post-parsing validators? initial assumption is: no

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes

Pull request closes #2688

This is a cleanup for section 2 validators. Required changes are listed [here](https://gorafttech.sharepoint.com/:x:/r/sites/TDRSResearchDesign/_layouts/15/Doc.aspx?sourcedoc=%7BFB4CFF15-BE50-4BF8-ADCC-7CC16A9AB128%7D&file=TDP_Validation_TANF_Section2.xlsx&action=default&mobileredirect=true)

## How to Test

These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
1. Upload test file
1. Confirm the fields as described in [here](https://gorafttech.sharepoint.com/:x:/r/sites/TDRSResearchDesign/_layouts/15/Doc.aspx?sourcedoc=%7BFB4CFF15-BE50-4BF8-ADCC-7CC16A9AB128%7D&file=TDP_Validation_TANF_Section2.xlsx&action=default&mobileredirect=true) can include space

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] space-fill accounted for tanf section 2
+ [x] sub housing range updated for tanf section 2
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/validators.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/validators.py]: TODO: have to append prefix zeros

**Comment by raftmsohani** [comment]
@ADPennington : The reference doc asks for accepting space filled race for T4, question is: do we wanna accept space filled for all other sections or only for T4?

File: tdrs-backend/tdpservice/parsers/validators.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/validators.py:244]: this is in fact not required but can improve the validator so instead of writing all values to be matched, we can add a range

File: tdrs-backend/tdpservice/parsers/test/data/small_tanf_section2.txt
Comment (elipe17) [inline tdrs-backend/tdpservice/parsers/test/data/small_tanf_section2.txt:2]: What's the change on this? Is it whitespace removal?

File: tdrs-backend/tdpservice/parsers/test/test_parse.py
Comment (elipe17) [inline tdrs-backend/tdpservice/parsers/test/test_parse.py]: Should the name of this be `test_empty_t4_t5_values` instead of `test_empty_t3_t5_values`?

File: tdrs-backend/tdpservice/parsers/test/data/small_tanf_section2.txt
Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/test/data/small_tanf_section2.txt:2]: I think the test below answered my question

**Review by elipe17 — APPROVED** [review]
Looks good! Just one question that doesn't block this from moving up in the review chain.

**Comment by raftmsohani** [comment]
> Looks good! Just one question that doesn't block this from moving up in the review chain.

Good catch! corrected

File: tdrs-backend/tdpservice/parsers/validators.py
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/validators.py:244]: could you include tests in `test_validators` for this change?

**Review by jtimpe — APPROVED** [review]
requested additional test, otherwise looks good!

**Comment by ADPennington** [comment]
unit tests need review here @raftmsohani

**Comment by raftmsohani** [comment]
> unit tests need review here @raftmsohani

sorry this happened after resolving merge conflict, resolving now

**Comment by raftmsohani** [comment]
> > unit tests need review here @raftmsohani
> 
> sorry this happened after resolving merge conflict, resolving now

@ADPennington  fixed!

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py]: @raftmsohani 3 is a valid possible response for this tribal field. 
2688p1

Comment (raftmsohani) [reply tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py]: @ADPennington I did this change based on [this](https://gorafttech.sharepoint.com/:x:/r/sites/TDRSResearchDesign/_layouts/15/Doc.aspx?sourcedoc=%7Bfb4cff15-be50-4bf8-adcc-7cc16a9ab128%7D&action=edit&wdinitialsession=d69272d6-fa55-5847-dcbc-6ff14510cbc4&wdrldsc=13&wdrldc=2&wdrldr=BootTimeMismatch). But I see your point, should I revert this back?

Comment (ADPennington) [reply tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py]: yes. the link points to the tanf closed workbook; not the tribal tanf closed workbook.

Comment (raftmsohani) [reply tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py]: Yes I checked it with [this](https://gorafttech.sharepoint.com/:x:/r/sites/TDRSResearchDesign/_layouts/15/Doc.aspx?sourcedoc=%7BCA4A37A5-64F7-421C-95DE-6D5970B0BFCD%7D&file=TDP_Validation_TribalTANF_Section2.xlsx&action=default&mobileredirect=true&DefaultItemOpen=1). However, I see you only have comment and did not ask to change it yet

Comment (ADPennington) [reply tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py]: k you can revert any changes to tribal tanf schema. we can revisit those in a separate ticket if these are needed. i havent observed any recent evidence of needing these changes. @raftmsohani

**Review by ADPennington — CHANGES_REQUESTED** [review]
this is in good shape. please revert the tribal tanf code changes.

**Review by ADPennington — APPROVED** [review]
lgtm @raftmsohani 🚀

**Pull request merged**
